File: Assets/Script/Test/VGButton.lua
-- [tsx]: VGButton.tsx
local ____exports = {} -- 1
local ____DoraX = require("DoraX") -- 2
local React = ____DoraX.React -- 2
local toNode = ____DoraX.toNode -- 2
local ____Dora = require("Dora") -- 3
local Color = ____Dora.Color -- 3
local Node = ____Dora.Node -- 3
local Size = ____Dora.Size -- 3
local nvg = require("nvg") -- 4
local function Button(props) -- 12
	local fontId = nvg.CreateFont("sarasa-mono-sc-regular") -- 13
	local light = nvg.LinearGradient( -- 14
		0, -- 14
		80, -- 14
		0, -- 14
		0, -- 14
		Color(4294967295), -- 14
		Color(4278255615) -- 14
	) -- 14
	local dark = nvg.LinearGradient( -- 15
		0, -- 15
		80, -- 15
		0, -- 15
		0, -- 15
		Color(4294967295), -- 15
		Color(4294689792) -- 15
	) -- 15
	local paint = light -- 16
	local function onCreate() -- 17
		local node = Node() -- 18
		node.size = Size(100, 100) -- 19
		node:onRender(function() -- 20
			nvg.ApplyTransform(node) -- 21
			nvg.BeginPath() -- 22
			nvg.RoundedRect( -- 23
				0, -- 23
				0, -- 23
				100, -- 23
				100, -- 23
				10 -- 23
			) -- 23
			nvg.StrokeColor(Color(4294967295)) -- 24
			nvg.StrokeWidth(5) -- 25
			nvg.Stroke() -- 26
			nvg.FillPaint(paint) -- 27
			nvg.Fill() -- 28
			nvg.ClosePath() -- 29
			nvg.FontFaceId(fontId) -- 30
			nvg.FontSize(32) -- 31
			nvg.FillColor(Color(4278190080)) -- 32
			nvg.Scale(1, -1) -- 33
			nvg.Text(50, -30, props.text) -- 34
			return false -- 35
		end) -- 20
		return node -- 37
	end -- 17
	return React.createElement( -- 39
		"custom-node", -- 39
		{ -- 39
			onCreate = onCreate, -- 39
			onTapBegan = function() -- 39
				paint = dark -- 42
				return paint -- 42
			end, -- 42
			onTapEnded = function() -- 42
				paint = light -- 43
				return paint -- 43
			end, -- 43
			onTapped = props.onClick, -- 43
			children = props.children -- 43
		} -- 43
	) -- 43
end -- 12
toNode(React.createElement( -- 50
	Button, -- 51
	{ -- 51
		text = "OK", -- 51
		onClick = function() return print("Clicked") end -- 51
	}, -- 51
	React.createElement( -- 51
		"sequence", -- 51
		nil, -- 51
		React.createElement("move-x", {time = 1, start = 0, stop = 200}), -- 51
		React.createElement("angle", {time = 1, start = 0, stop = 360}), -- 51
		React.createElement("scale", {time = 1, start = 1, stop = 4}) -- 51
	) -- 51
)) -- 51
return ____exports -- 51

File: Tools/dora-rust/dora-test/src/tests/vg_button.rs
/* Copyright (c) 2016-2026 Li Jin <dragon-fly@qq.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. */

use dora_ssr::*;
use std::{cell::RefCell, rc::Rc};

pub fn test() {
	let mut node = Node::new();
	node.set_size(&Size::new(100.0, 100.0));
	let font_id = Nvg::create_font("sarasa-mono-sc-regular");
	let light = Rc::new(Nvg::linear_gradient(
		0.0, 60.0, 0.0, 0.0,
		&Color::new(0xffffffff),
		&Color::new(0xff00ffff)
	));
	let dark = Rc::new(Nvg::linear_gradient(
		0.0, 60.0, 0.0, 0.0,
		&Color::new(0xffffffff), &Color::new(0xfffbc400)
	));
	let paint = Rc::new(RefCell::new(light.clone()));
	let node_clone = node.clone();
	let paint_clone = paint.clone();
	node.on_render(Box::new(move |_| {
		Nvg::apply_transform(&node_clone);
		Nvg::begin_path();
		Nvg::rounded_rect(0.0, 0.0, 100.0, 100.0, 10.0);
		Nvg::stroke_color(&Color::new(0xffffffff));
		Nvg::stroke_width(5.0);
		Nvg::stroke();
		Nvg::fill_paint(&paint_clone.borrow());
		Nvg::fill();
		Nvg::close_path();
		Nvg::font_face_id(font_id);
		Nvg::font_size(32.0);
		Nvg::fill_color(&Color::new(0xff000000));
		Nvg::scale(1.0, -1.0);
		Nvg::text(50.0, -30.0, "OK");
		false
	}));
	node.perform_def(ActionDef::sequence(&vec![
		ActionDef::prop(1.0, 0.0, 200.0, Property::X, EaseType::Linear),
		ActionDef::prop(1.0, 0.0, 360.0, Property::Angle, EaseType::Linear),
		ActionDef::scale(1.0, 1.0, 4.0, EaseType::Linear)
	]), false);
	let paint_clone = paint.clone();
	Slot::on_tap_began(&mut node, move |_| {
		*paint_clone.borrow_mut() = dark.clone();
	});
	let paint_clone = paint.clone();
	Slot::on_tap_ended(&mut node, move |_| {
		*paint_clone.borrow_mut() = light.clone();
	});
	Slot::on_tapped(&mut node, move |_| {
		p!("Clicked!");
	});
}